src/internal/pfsdb/commits.go

```diff
@@ -0,0 +1,994 @@
+}
+
+// AncestryOpt allows users to create commitInfos and skip creating the ancestry information.
+// This allows a user to create the commits in an arbitrary order, then create their ancestry later.
+type AncestryOpt struct {
+	SkipChildren bool
+	SkipParent   bool
+}
+
+// CreateCommit creates an entry in the pfs.commits table. If the commit has a parent or children,
+// it will attempt to create entries in the pfs.commit_ancestry table unless options are provided to skip
+// ancestry creation.
+func CreateCommit(ctx context.Context, tx *pachsql.Tx, commitInfo *pfs.CommitInfo, opts ...AncestryOpt) (CommitID, error) {
+	if err := validateCommitInfo(commitInfo); err != nil {
+		return 0, err
+	}
+	opt := AncestryOpt{}
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+	insert := Commit{
+		CommitID:    CommitKey(commitInfo.Commit),
+		CommitSetID: commitInfo.Commit.Id,
+		Repo: Repo{
+			Name: commitInfo.Commit.Repo.Name,
+			Type: commitInfo.Commit.Repo.Type,
+			Project: coredb.Project{
+				Name: commitInfo.Commit.Repo.Project.Name,
+			},
+		},
+		BranchName:     sql.NullString{String: commitInfo.Commit.Branch.Name, Valid: true},
+		Origin:         commitInfo.Origin.Kind.String(),
+		StartTime:      sanitizeTimestamppb(commitInfo.Started),
+		FinishingTime:  sanitizeTimestamppb(commitInfo.Finishing),
+		FinishedTime:   sanitizeTimestamppb(commitInfo.Finished),
+		Description:    commitInfo.Description,
+		CompactingTime: durationpbToBigInt(commitInfo.Details.CompactingTime),
+		ValidatingTime: durationpbToBigInt(commitInfo.Details.ValidatingTime),
+		Size:           commitInfo.Details.SizeBytes,
+		Error:          commitInfo.Error,
+	}
+	namedStmt, err := tx.PrepareNamedContext(ctx, createCommit)
+	if err != nil {
+		return 0, errors.Wrap(err, "prepare create commitInfo")
+	}
+	row := namedStmt.QueryRowxContext(ctx, insert)
+	if row.Err() != nil {
+		if IsDuplicateKeyErr(row.Err()) { // a duplicate key implies that an entry for the repo already exists.
+			return 0, ErrCommitAlreadyExists{CommitID: CommitKey(commitInfo.Commit)}
+		}
+		return 0, errors.Wrap(row.Err(), "exec create commitInfo")
+	}
+	lastInsertId := 0
+	if err := row.Scan(&lastInsertId); err != nil {
+		return 0, errors.Wrap(err, "scanning id from create commitInfo")
+	}
+	if commitInfo.ParentCommit != nil && !opt.SkipParent {
+		if err := CreateCommitParent(ctx, tx, commitInfo.ParentCommit, CommitID(lastInsertId)); err != nil {
+			return 0, errors.Wrap(err, "linking parent")
+		}
+	}
+	if len(commitInfo.ChildCommits) != 0 && !opt.SkipChildren {
+		if err := CreateCommitChildren(ctx, tx, CommitID(lastInsertId), commitInfo.ChildCommits); err != nil {
+			return 0, errors.Wrap(err, "linking children")
+		}
+	}
+	return CommitID(lastInsertId), nil
+}
+
+// CreateCommitParent inserts a single ancestry relationship where the child is known and parent must be derived.
+func CreateCommitParent(ctx context.Context, tx *pachsql.Tx, parentCommit *pfs.Commit, childCommit CommitID) error {
```

Comment:
Hmm I see you have 5 different flavours of creating an edge in the `pfs.commit_ancestry` table. I think that's a bit excessive.

How about we have something like

```Go
type ParentChild struct {
    parent *Commit
    child *Commit
}

func CreateCommitAncestry(ctx context.Context, tx *pachsql.Tx, ancestry ParentChild) error

func CreateCommitAncestryBatch(ctx context.Context, tx *pachsql.Tx, ancestries []ParentChild) error

// CreateCommitAncestryByName will resolve the commit ids internally and use CreateCommitAncestry
func CreateCommitAncestryByName(ctx context.Context, tx *pachsql.Tx, parent, child *pfs.Commit) error
```